trio/_util.py

```diff
@@ -242,3 +245,31 @@ def fspath(path) -> t.Union[str, bytes]:
 
 if hasattr(os, "fspath"):
     fspath = os.fspath  # noqa
+
+
+class generic_function:
+    """Decorator that makes a function indexable, to communicate
+    non-inferrable generic type parameters to a static type checker.
+
+    If you write::
+
+        @generic_function
+        def open_memory_channel(max_buffer_size: int) -> Tuple[
+            SendChannel[T], ReceiveChannel[T]
+        ]: ...
+
+    it is valid at runtime to say ``open_memory_channel[bytes](5)``.
+    This behaves identically to ``open_memory_channel(5)`` at runtime,
+    and currently won't type-check without a mypy plugin or clever stubs,
+    but at least it becomes possible to write those.
+    """
+
+    def __init__(self, fn):
+        update_wrapper(self, fn)
+        self._fn = fn
+
+    def __call__(self, *args, **kwargs):
+        return self._fn(*args, **kwargs)
+
+    def __getitem__(self, _):
+        return self
```

Comment:
So you actually can make `open_memory_channel` directly typeable by mypy without any plugin, and it's not too terribly awful: https://github.com/python/mypy/issues/6073#issuecomment-447616140

But I think it does require inlining this class into the definition of `open_memory_channel`.

If that's where were going to end up, then maybe setting up a standalone `@generic_function` decorator, testing it, etc., is overkill, and we should instead move this code into `_memory_channels.py` right now?

Reply:
The problem with that approach is that AFAICT, when you write `Type[T]`, mypy can't infer `T` as one of its "special forms" (Tuple, Union, Callable, etc). You just silently get `Any` if you use such a one. I figured that was error-prone enough that we should go this route instead, at least for now.

Reply:
Oh ick.